File: lecture04/02_happy/template.cpp
// happy_number.cpp

#include <iostream>
#include <numeric>
#include <unordered_map>
#include <unordered_set>

using namespace std;

// Global variables

// Functions

bool is_happy_r(int n, unordered_set<int> &seen) {
    // TODO: Determine if n is a happy number (recursively)
    return false;
}

bool is_happy(int n) {
    // TODO: Determine if n is a happy number
    return false;
}

// Main execution

int main(int argc, char *argv[]) {
    int n;

    while (cin >> n) {
    	cout << (is_happy(n) ? "Yes" : "No") << endl;
    }

    return 0;
}

File: lecture04/02_happy/template.py
#!/usr/bin/env python3

import sys

# Naive version

def is_happy(n, seen):
    # TODO: Determine if n is a happy number (recursively)
    return False

# Main execution

for n in map(int, sys.stdin):
    print('Yes' if is_happy_cached(n, set()) else 'No')
